crates/proto/src/xfer/dnssec_dns_handle.rs

```diff
@@ -1022,6 +1043,726 @@ fn current_time() -> u32 {
+            .iter()
+            .find(|record| record.base32_hashed_name == base32_hashed_query_name.as_bytes());
+
+        // Case 2:
+        // Name exists but there's no record of this type
+        if let Some(query_record) = query_name_record {
+            if query_record
+                .nsec3_data
+                .type_bit_maps()
+                .contains(&query_type)
+            {
+                return Proof::Bogus;
+            } else {
+                return Proof::Secure;
+            }
+        }
+
+        match wildcard_encloser_num_labels {
+            // Case 3:
+            // Name is covered by wildcard that has a record of this type
+            Some(wildcard_encloser_num_labels) => {
+                if query_name.num_labels() <= wildcard_encloser_num_labels {
+                    return Proof::Bogus;
+                }
+                let next_closer_labels = query_name
+                    .into_iter()
+                    .rev()
+                    .take(wildcard_encloser_num_labels as usize + 1)
+                    .rev()
+                    .collect::<Vec<_>>();
+                let next_closer_name = Name::from_labels(next_closer_labels).unwrap();
+                let next_closer_hashed_name = nsec3hash(&next_closer_name, salt, iterations);
+                let next_closer_base32_hashed_name =
+                    data_encoding::BASE32_DNSSEC.encode(&next_closer_hashed_name);
+                let next_closer_record = find_covering_record(
+                    nsec3s,
+                    &next_closer_hashed_name,
+                    &next_closer_base32_hashed_name,
+                );
+                match next_closer_record {
+                    Some(_) => Proof::Secure,
+                    None => Proof::Bogus,
+                }
+            }
+
+            // Case 4:
+            // Name is covered by wildcard that doesn't have a record of this type
+            None => {
+                let ClosestEncloserProofInfo {
+                    closest_encloser,
+                    next_closer,
+                    closest_encloser_wildcard,
+                } = wildcard_based_encloser_proof(query_name, soa_name, nsec3s);
+                match (closest_encloser, next_closer, closest_encloser_wildcard) {
+                    (Some(_), Some(_), Some(_)) => Proof::Secure,
+                    (None, Some(_), Some(_)) if &query_name.base_name() == soa_name => {
+                        Proof::Secure
+                    }
+                    (None, None, None) if query_name == soa_name => Proof::Secure,
+                    _ => Proof::Bogus,
+                }
+            }
+        }
+    }
+
+    // Unlick non-wildcard version this cannot produce the early `Proof``
+    fn wildcard_based_encloser_proof<'a>(
+        query_name: &Name,
+        soa_name: &Name,
+        nsec3s: &'a [Nsec3RecordPair<'a>],
+    ) -> ClosestEncloserProofInfo<'a> {
+        let salt = nsec3s[0].nsec3_data.salt();
```

Comment:
How is the invariant upheld that `nsec3s` is non-empty?

Reply:
These functions are only called from main `verify_nsec3` where I specifically check for that. Also, the main `verify_nsec3` function is only called when there are `NSEC3` records present.

Would you still like to see a debug assert here?

Reply:
Yes, I think a `debug_assert!()` would make sense.